src/Transports.AspNetCore/GraphQLHttpMiddleware.cs

```diff
@@ -108,219 +287,398 @@ public virtual async Task InvokeAsync(HttpContext context, RequestDelegate next)
-        GraphQLRequest gqlRequest,
-        IDocumentExecuter<TSchema> executer,
-        CancellationToken cancellationToken)
+        GraphQLRequest gqlRequest)
     {
         // Normal execution with single graphql request
-        if (bodyGQLBatchRequest == null)
-        {
-            var stopwatch = ValueStopwatch.StartNew();
-            await RequestExecutingAsync(gqlRequest);
-            var result = await ExecuteRequestAsync(gqlRequest, userContext, executer, context.RequestServices, cancellationToken);
-
-            await RequestExecutedAsync(new GraphQLRequestExecutionResult(gqlRequest, result, stopwatch.Elapsed));
+        var userContext = await BuildUserContextAsync(context);
+        var result = await ExecuteRequestAsync(context, gqlRequest, context.RequestServices, userContext);
+        var statusCode = Options.ValidationErrorsReturnBadRequest && !result.Executed
+            ? HttpStatusCode.BadRequest
+            : HttpStatusCode.OK;
+        await WriteJsonResponseAsync(context, statusCode, result);
+    }
 
-            await WriteResponseAsync(context.Response, _serializer, cancellationToken, result);
+    /// <summary>
+    /// Handles a batched GraphQL request.
+    /// </summary>
+    protected virtual async Task HandleBatchRequestAsync(
+        HttpContext context,
+        RequestDelegate next,
+        IList<GraphQLRequest?> gqlRequests)
+    {
+        var userContext = await BuildUserContextAsync(context);
+        var results = new ExecutionResult[gqlRequests.Count];
+        if (gqlRequests.Count == 1)
+        {
+            results[0] = await ExecuteRequestAsync(context, gqlRequests[0], context.RequestServices, userContext);
         }
-        // Execute multiple graphql requests in one batch
         else
         {
-            var executionResults = new ExecutionResult[bodyGQLBatchRequest.Count];
-            for (int i = 0; i < bodyGQLBatchRequest.Count; ++i)
+            // Batched execution with multiple graphql requests
+            if (!Options.ExecuteBatchedRequestsInParallel)
             {
-                var gqlRequestInBatch = bodyGQLBatchRequest[i];
-
-                var stopwatch = ValueStopwatch.StartNew();
-                await RequestExecutingAsync(gqlRequestInBatch, i);
-                var result = await ExecuteRequestAsync(gqlRequestInBatch, userContext, executer, context.RequestServices, cancellationToken);
-
-                await RequestExecutedAsync(new GraphQLRequestExecutionResult(gqlRequestInBatch, result, stopwatch.Elapsed, i));
-
-                executionResults[i] = result;
+                for (int i = 0; i < gqlRequests.Count; i++)
+                {
+                    results[i] = await ExecuteRequestAsync(context, gqlRequests[i], context.RequestServices, userContext);
+                }
+            }
+            else
+            {
+                var resultTasks = new Task<ExecutionResult>[gqlRequests.Count];
```

Comment:
```suggestion
                // TODO: implement some sort of concurrency limit
                var resultTasks = new Task<ExecutionResult>[gqlRequests.Count];
```

Reply:
I wouldn't do it at all, and we already have an issue for it if someone wants to write a PR.  There is no reason to add a comment in the code on top.

Reply:
#793